### PR TITLE
[libxenbackend] Check val before using it

### PR DESCRIPTION
**Pull request by crogers1**
It's possible in certain cases for the xs_read for val to return
  null if the otherend closed its xenstore node before it could be
  read.  Make sure we check 'val' before using or freeing it to
  avoid segfaulting.

Signed-off-by: Chris Rogers <crogers122@gmail.com>

### CODE REVIEW

File: src/state.c [hunk 1]
```diff
@@ -57,11 +57,11 @@ void backend_changed(struct xen_device *xendev, const char *node)
     if (node) {
         char *val = xs_read_be_str(xendev, node);
 
-        if (xenback->ops->backend_changed)
+        if (xenback->ops->backend_changed && val)
```
Comment (jandryuk) [inline src/state.c]: I think we may still want to call backend_changed, even if val is NULL.  If the path is deleted, the watch will fire and val will be NULL.  We probably want to still call backend_changed in that case, and that callback should be ready for NULL.  Silently dropping a delete doesn't seem right to me.

Having said that, I think vGlass is the only user of libxenbackend now, and it doesn't defined a backend_changed callback.  So nothing is using the callback.

File: src/state.c [hunk 2]
```diff
@@ -57,11 +57,11 @@ void backend_changed(struct xen_device *xendev, const char *node)
...
             xenback->ops->backend_changed(xendev->dev, node, val);
 
         /* We were told to close */
-        if (strcmp(node, "state") == 0 && strcmp(val, "5") == 0) {
+        if (strcmp(node, "state") == 0 && val && strcmp(val, "5") == 0) {
```
Comment (jandryuk) [inline src/state.c:64]: This is definitely needed since it is looking at the value.

File: src/state.c [hunk 3]
```diff
@@ -70,7 +70,8 @@ void backend_changed(struct xen_device *xendev, const char *node)
             }
         }
 
-        free(val);
+        if (val)
+            free(val);
```
Comment (jandryuk) [inline src/state.c]: `free()` can take a NULL pointer, so this isn't really needed.

**Review by jandryuk — COMMENTED** [review]
Thoughts on not adding the NULL check before backend_changed?

File: src/state.c (hunk 3 above)
Comment (crogers1) [reply src/state.c]: agreed

File: src/state.c (hunk 1 above)
Comment (crogers1) [reply src/state.c]: Also agreed

**Review by jandryuk — APPROVED** [review]

**Review by jandryuk — APPROVED** [review]
LGTM - Thanks

**Comment by jandryuk** [comment]
Will merge soon.

**Pull request merged**